### PR TITLE
CommitteeState query

### LINKED ISSUE #3750: Replace CommitteeState with CommitteMembersState query
Current implementation of `queryCommitteeState` is confusing at best, but at the worst provides incomplete story. There is also a place in the EnactState where store the actual cold keys for the current committee and their respective expiration epoch numbers.

https://github.com/input-output-hk/cardano-ledger/blob/15ccf21728173060576aca9799c06eef24b143df/libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs#L121-L123

We need to change this query to this semantics:

```haskell

-- | Query committee members
queryCommitteeMembersState ::
  Set (Credential 'ColdCommitteeRole c) ->
  -- ^ When empty, do not filter by the cold credential, otherwise return only members
  -- that are in this set
  Set (Credential 'HotCommitteeRole c) ->
  -- ^ When empty, do not filter by the cold credential, otherwise return only members
  -- that are in this set
  Maybe MemberStatus ->
  -- ^ Only return members with this status. Useful for discovering only active members,
  -- for example. When Nothing no filtering by status will occur.
  NewEpochState era ->
  CommitteeMembersState era
queryCommitteeMembersState = ...

data HotCredAuthStatus c
  = MemberHaveNotAuthorized
  -- ^ Member have been Enacted, but no hot credential for voting have been authorized
  | MemberResigned
  | Authorized (Credential 'HotCommitteeRole c)

data MemberStatus
  = -- | This can happen when someone has authorized a hot credential for an unknown cold
    -- credential. Such Committee member will either be removed from the state at the next
    -- epoch boundary or be enacted as a new member.
    Unrecognized
  | -- | This is an expired member
    Expired
  | -- | VOtes of this Commitee member wi
    Active

data NextEpochChange
  = -- | This committee member has not been enacted yet, but it will be at the next epoch.
    ToBeEnacted
  | -- | This committee member has will be removed from the state. Either because it wa
    ToBeRemoved
  | -- | Nothing expected to change for the committee member
    NoChangeExpected

data CommitteeMembersState c = CommitteeMemberState
  { cmsHotCredAuthStatus :: HotCredAuthStatus c
  , cmsStatus :: MemberStatus
  , cmsExpiration :: Maybe EpochNo
  -- ^ Absolute epoch number when the member will or have expired
  , csNextEpochChange :: NextEpochChange
  -- ^ Are there any changes planned for this Constitutional Committee member at the next
  -- epoch
  }

data CommitteeState c = CommitteState
  { csCommittee :: Map (Credential 'ColdCommitteeRole c) (CommitteeMemberState c)
  , csEpochNo :: EpochNo
  -- ^ Current epoch number. This is necessary to interpret states of some of the
  -- Committee members
  }
```

Result type will need a stable ToCBOR/FromCBOR instances as well as To/FromJSON instances.

### PR DESCRIPTION
**Pull request by teodanciu**
# Description

Closes https://github.com/input-output-hk/cardano-ledger/issues/3750 

~Still draft because I"m uncertain about a few things, to be discussed tomorrow.~

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs [hunk 1]
```diff
@@ -16,3 +44,212 @@ spec = do
...
+        Map.fromList $
+          zipWith
+            (\(k1, _) (_, v2) -> (k1, v2))
+            (Map.assocs chosen)
+            (Map.assocs comStateMembers)
+  pure $ CommitteeState $ x `Map.union` (Map.drop s comStateMembers)
+
+propEmpty :: NewEpochState Conway -> Expectation
+propEmpty nes =
+  do
```
Comment (lehins) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs]: Could you please avoid this style?
Makes it look like we are using 4 space indentation
```suggestion
propEmpty nes = do
```

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs [hunk 2]
```diff
@@ -16,3 +44,212 @@ spec = do
...
+    let (comMembers, comStateMembers) = committeeInfo nes
+    let (CommitteeMembersState result _) = queryCommitteeMembersStateNoFilters nes
```
Comment (lehins) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs]: Redundant parenthesis
```suggestion
    let CommitteeMembersState result _ = queryCommitteeMembersStateNoFilters nes
```

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs [hunk 3]
```diff
@@ -16,3 +44,212 @@ spec = do
...
+    let (_, comStateMembers) = committeeInfo nes
+    let (CommitteeMembersState result _) = queryCommitteeMembersStateNoFilters nes
+    let notAuthorized =
+          Map.filter
+            ( \case
+                CommitteeMemberState MemberNotAuthorized _ _ _ -> True
+                _ -> False
+            )
+            result
+    -- if the member is NotAuthorized, it should not have an associated hot credential in the commiteSTate
```
Comment (lehins) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs]:
```suggestion
    -- if the member is NotAuthorized, it should not have an associated hot credential in the commiteState
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+instance DecCBOR MemberStatus where
+  decCBOR = decodeEnumBounded
+
+instance ToCBOR MemberStatus where
+  -- toCBOR = toPlainEncoding (eraProtVerLow @Conway) . encCBOR
+  toCBOR = toEraCBOR @Conway
+
+instance FromCBOR MemberStatus where
+  -- fromCBOR = toPlainDecoder (eraProtVerLow @Conway) decCBOR
+  fromCBOR = fromEraCBOR @Conway
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs]: These instances can't exist, since there is no access to `era` parameter. At the same time they are not needed either. At the place where they will be serialized/deserialized `era` will be available and that is where protocol version can be inferred. Here is an example how it works:
https://github.com/input-output-hk/ouroboros-consensus/blob/903ba3c7d2f8d8c015b59f0c664cf75c0f280c25/ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Ledger/Query.hs#L812
```suggestion
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+      0 -> SumD MemberAuthorized <! From
+      1 -> SumD MemberNotAuthorized
+      2 -> SumD MemberResigned
+      k -> Invalid k
+
+instance Crypto c => ToCBOR (HotCredAuthStatus c) where
+  toCBOR = toEraCBOR @Conway
+
+instance Crypto c => FromCBOR (HotCredAuthStatus c) where
+  fromCBOR = fromEraCBOR @Conway
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs]: Same here
```suggestion
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        <! From
+        <! From
+        <! From
+        <! From
+
+instance Crypto c => ToCBOR (CommitteeMemberState c) where
+  toCBOR = toEraCBOR @Conway
+
+instance Crypto c => FromCBOR (CommitteeMemberState c) where
+  fromCBOR = fromEraCBOR @Conway
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs]: Ditto
```suggestion
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    decode $
+      RecD CommitteeMembersState
+        <! From
+        <! From
+
+instance Crypto c => ToCBOR (CommitteeMembersState c) where
+  toCBOR = toEraCBOR @Conway
+
+instance Crypto c => FromCBOR (CommitteeMembersState c) where
+  fromCBOR = fromEraCBOR @Conway
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs]: Ditto
```suggestion
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+  encCBOR = encodeEnum
+
+instance DecCBOR NextEpochChange where
+  decCBOR = decodeEnumBounded
+
+instance ToCBOR NextEpochChange where
+  toCBOR = toEraCBOR @Conway
+
+instance FromCBOR NextEpochChange where
+  fromCBOR = fromEraCBOR @Conway
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs]: Ditto
```suggestion
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+      , "expiration" .= cmsExpiration
+      , "nextEpochChange" .= cmsNextEpochChange
+      ]
+
+data CommitteeMembersState c = CommitteeMembersState
+  { csCommittee ::
+      Map
+        (Credential 'ColdCommitteeRole c)
+        (CommitteeMemberState c)
+  , csEpochNo :: EpochNo
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs]: Let's make 'em strict and this line isn't too long yet
```suggestion
  { csCommittee :: !(Map (Credential 'ColdCommitteeRole c) (CommitteeMemberState c))
  , csEpochNo :: !EpochNo
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+instance FromCBOR NextEpochChange where
+  fromCBOR = fromEraCBOR @Conway
+
+data CommitteeMemberState c = CommitteeMemberState
+  { cmsHotCredAuthStatus :: HotCredAuthStatus c
+  , cmsStatus :: MemberStatus
+  , cmsExpiration :: Maybe EpochNo
+  -- ^ Absolute epoch number when the member expires
+  , cmsNextEpochChange :: NextEpochChange
+  -- ^ Changes to the member at the next epoch
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/CommitteeMembersState.hs]:
```suggestion
  { cmsHotCredAuthStatus :: !(HotCredAuthStatus c)
  , cmsStatus :: !MemberStatus
  , cmsExpiration :: !(Maybe EpochNo)
  -- ^ Absolute epoch number when the member expires
  , cmsNextEpochChange :: !NextEpochChange
  -- ^ Changes to the member at the next epoch
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs [hunk 11]
```diff
@@ -121,3 +143,108 @@ queryDRepStakeDistr nes creds
...
+          else comMembers `Map.restrictKeys` coldCredsFilter
+
+      -- Start from the committee and lookup the hot credential in the Committee State.
+      -- This will miss 'Unrecognized' members (since we're iterating over the "recognized" ones).
+      recognized :: [(Credential 'ColdCommitteeRole (EraCrypto era), CommitteeMemberState (EraCrypto era))]
+      recognized =
+        Map.assocs filteredComMembers >>= \(ck, expirationEpoch) ->
+          let mbHk = Map.lookup ck comStateMembers
+              mbCommitteeMemberState =
+                do
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs]:
```suggestion
              mbCommitteeMemberState = do
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs [hunk 12]
```diff
@@ -121,3 +143,108 @@ queryDRepStakeDistr nes creds
...
+                      CommitteeMemberState
+                        hkStatus
+                        Unrecognized
+                        Nothing
+                        nextEpochChange
+                pure (ck, cms)
+           in maybeToList mbCommitteeMemberState
+   in CommitteeMembersState
+        { csCommittee = Map.fromList (recognized ++ unrecognized)
+        , csEpochNo = nes ^. nesELL - 1
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs]: Why is this change to previous epoch? This query suppose to be for the current state.
```suggestion
        , csEpochNo = nes ^. nesELL
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs [hunk 13]
```diff
@@ -121,3 +143,108 @@ queryDRepStakeDistr nes creds
...
+                  pure (ck, cms)
+           in maybeToList mbCommitteeMemberState
+
+      filteredUnrecognizedMembers =
+        if null coldCredsFilter
+          then Map.difference comStateMembers comMembers
+          else
+            Map.restrictKeys
+              (Map.difference comStateMembers comMembers)
+              coldCredsFilter
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs]: Slight reduction in duplication. Also usage of guards as in binding below
```suggestion
      unrecognizedMembers = Map.difference comStateMembers comMembers
      filteredUnrecognizedMembers
        | null coldCredsFilter = unrecognizedMembers
        | otherwise = Map.restrictKeys unrecognizedMembers coldCredsFilter
```

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs [hunk 14]
```diff
@@ -121,3 +143,108 @@ queryDRepStakeDistr nes creds
...
+  -- | filter by cold credentials (don't filter when empty)
+  Set (Credential 'ColdCommitteeRole (EraCrypto era)) ->
+  -- | filter by hot credentials (don't filter when empty)
+  Set (Credential 'HotCommitteeRole (EraCrypto era)) ->
+  -- | filter by status (don't filter when empty)
+  -- (useful, for discovering, for example, only active members)
+  Set MemberStatus ->
+  NewEpochState era ->
+  CommitteeMembersState (EraCrypto era)
+queryCommitteeMembersState coldCredsFilter hotCredsFilter statusFilter nes =
```
Comment (lehins) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs:165]: Current implementation of this query is a bit complicated and it is hard to follow. I suggest that it will be easier to start with a set of relevant cold keys and convert it to a Map with some filtering, something along these lines:
```haskell

      relevantHotCreds :: Set (Credential 'ColdCommitteeRole (EraCrypto era))
      relevantHotCreds
        | null hotCredsFilter = comStateMembers
        | otherwise = Map.filter (maybe False (`Set.member` hotCredsFilter)) comStateMembers

      relevantColdCreds :: Set (Credential 'ColdCommitteeRole (EraCrypto era))
      relevantColdCreds
        -- we only look for Unrecognized members when it is contained in the filter, or when there is no filtering
        | Set.null statusFilter || Set.member Unrecognized statusFilter =
            Map.keysSet comMembers `Set.union` Map.keysSet relevantHotCredsMap
        | otherwise = Map.keysSet comMembers

      filteredColdCreds
        | null coldCredsFilter = relevantColdCreds
        | otherwise = relevantColdCreds `Map.restrictKeys` coldCredsFilter

      committeMemeberState = 
        Map.mapMaybe id $ Map.fromSet mkMaybeMemberState filteredColdCreds

      mkMaybeMemberState :: Credential 'ColdCommitteeRole (EraCrypto era) ->
                            Maybe (CommitteeMemberState (EraCrypto era))
      mkMaybeMemberState coldCred = do
        let status =
              case Map.lookup coldCred comMembers of
                Nothing -> Unrecognized
                Just expiry | nes ^. nesELL > expiry -> Expired
                            | otherwise -> Active
        guard (null statusFilter || status `Set.member` statusFilter)
        ...
```
This way you can focus on a single cold credential at a time in `mkMaybeMemberState`, which only needs to filter by status and you can use `do` syntax for Maybe to keep it readable.

**Review by lehins — CHANGES_REQUESTED** [review]
> I went down the wrong path

No worries, it happens.

This PR is on the right path :wink:

File: libs/cardano-ledger-api/cardano-ledger-api.cabal [hunk 15]
```diff
@@ -36,6 +36,7 @@ library
         Cardano.Ledger.Api.Tx.In
         Cardano.Ledger.Api.Tx.Out
         Cardano.Ledger.Api.Tx.Wits
+        Cardano.Ledger.Api.State.CommitteeMembersState
```
Comment (lehins) [inline libs/cardano-ledger-api/cardano-ledger-api.cabal]: This should be a hidden module with all types re-exported from `Cardano.Ledger.Api.State.Query`

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs (hunk 12 above)
Comment (teodanciu) [reply libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs]: I wrote it like this because of [how we create the RatifyEnv in Ledger](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs#L300), by passing it `currentEpoch - 1`, but I suppose, from the perspective of the caller of the Query, the epochNo is the one from the LedgerState.

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs (hunk 14 above)
Comment (teodanciu) [reply libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs:165]: Nice suggestion, thank you. 
I shuffled it around, because in this form it's  a little wrong (the `union` after the `filter` overwrites the filter), but I kept the same idea.

File: libs/cardano-ledger-api/cardano-ledger-api.cabal [hunk 16]
```diff
@@ -40,7 +40,10 @@ library
         Cardano.Ledger.Api.UTxO
 
     hs-source-dirs:   src
-    other-modules:    Cardano.Ledger.Api.Scripts.ExUnits
+    other-modules:
+        Cardano.Ledger.Api.Scripts.ExUnits
+        Cardano.Ledger.Api.State.CommitteeMembersState
```
Comment (lehins) [inline libs/cardano-ledger-api/cardano-ledger-api.cabal]: Could you move this module into a Query subfolder, since we'll have a lot more State, but not query related modules in the `Cardano/Ledger/Api/State` path
```suggestion
        Cardano.Ledger.Api.State.Query.CommitteeMembersState
```

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs [hunk 17]
```diff
@@ -16,3 +45,226 @@ spec = do
...
+
+queryCommitteeMembersStateNoFilters :: NewEpochState Conway -> CommitteeMembersState StandardCrypto
+queryCommitteeMembersStateNoFilters =
+  queryCommitteeMembersState @Conway
+    Set.empty
+    Set.empty
+    Set.empty
+
+instance Arbitrary MemberStatus where
+  arbitrary = arbitraryBoundedEnum
```
Comment (lehins) [inline libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs]: Can you add a module to the `testlib`: `Test.Cardano.Ledger.Api.Arbitrary` and move this instance in there.

**Review by lehins — APPROVED** [review]
Looks great. Thank you!

File: libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs (hunk 17 above)
Comment (teodanciu) [reply libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/State/QuerySpec.hs]: I actually did that at first, and when I saw it's only this that needs an instance ,I thought it's "not worth it", but I agree, I prefer it in a different module, even if it's two lines.  That's how big messes start, line by line :laughing: 
Moved it now, thanks for suggesting .

File: libs/cardano-ledger-api/cardano-ledger-api.cabal (hunk 16 above)
Comment (teodanciu) [reply libs/cardano-ledger-api/cardano-ledger-api.cabal]: Yes, sorry for forgetting, I know you told me yesterday. Done
